### PR TITLE
Ccdb fetcher

### PR DESCRIPTION
**Pull request by ktf**
Ongoing:

- [x] Understand why my changes to the CCDBApi seem to crash some other tests.
- [x] Add the fetching of the CTP and associated calculation of the timestamp based on OrbitResetTime and firstTFOrbit
- [x] Do not deserialise the objects every timeframe
- [x] Add a callback when it actually needs to happen.
- [x] Use DISTSUBTIMEFRAME / TF/TFNUM where appropriate to get the time for the data
- [x] Add ConfigParamSpec for createdNotAfter and createdNotBefore
- [x] Support for metadata on query in the InputSpec
- [x] Allow configuring the CCDB path / backend on a per input basis.
- [ ] Support for headers as result message.
- [ ] Better strategy to cleanup the cache and eventually support prefetching of CCDB objects / multiple in-fly objects for the same path.

### CODE REVIEW

**Comment by ktf** [comment]
First iteration, do not merge.

File: Framework/Core/src/CCDBHelpers.cxx
Comment (shahor02) [inline Framework/Core/src/CCDBHelpers.cxx]: I think `v.size()==0` should be treated as an error only if the ETAG was empty (cache was not filled yet), with ETAG for the valid object the `v` must be empty.

Comment (ktf) [reply Framework/Core/src/CCDBHelpers.cxx]: agreed.

**Comment by ktf** [comment]
With the latest commit commit the example in `Framework/TestWorkflows/src/test_CCDBFetcher.cxx` works as before when running with:

```
stage/bin/o2-testworkflows-test-ccdb-fetcher --start-value-enumeration `date +%s` --condition-backend http://alice-ccdb.cern.ch:8080
```

but using the new centralised logic and honouring the etag.

File: Framework/Core/src/CCDBHelpers.cxx
Comment (ktf) [reply Framework/Core/src/CCDBHelpers.cxx]: done.

**Comment by shahor02** [comment]
>> Understand why my changes to the CCDBApi seem to crash some other tests.

I think this is because of exposing boost headers to root cling. Perhaps we can avoid this by templating the vector to fill?

**Comment by ktf** [comment]
I was thinking about that, but then we would need to expose everyone a  bunch of ROOT headers. I am now trying something  slightly different with `ifndef __CLING__`.

**Comment by ktf** [comment]
New version fetches the OrbitResetTime from the CTP (plus a bit of cleanup).

**Comment by ktf** [comment]
Update now does not deserialise CCDB objects if the associated blob was not changed.

**Comment by ktf** [comment]
Now with the callback for deserialised objects.

File: Framework/Core/include/Framework/Task.h
Comment (shahor02) [inline Framework/Core/include/Framework/Task.h]: typo in "finaliseCCDB"?

Comment (ktf) [reply Framework/Core/include/Framework/Task.h]: yes

**Comment by ktf** [comment]
Fixed some tests and update to include DISTSUBTIMEFRAME / TF/TFNUM subscription, if present in the topology.

**Comment by ktf** [comment]
Metadata support added.

**Comment by ktf** [comment]
I now have the remapping in place. There is probably something still to adapt for the multihost case, but it should work for simple, same server remappings. The syntax is:

```
--condition-remap <prefix1>=<target1>,[<target2>;<prefix2>=<target3>[,...]]
```

where prefix can be either `https://` / `http://` or `/`. The targets are the paths which are specified by the `ccdbParamSpec()`. For the time being the must fully match.

If this works I would merge this PR and address the remaining issues in separate developments.

**Comment by shahor02** [comment]
> For the time being they must fully match.

If you refer to URLs with the wildcard, I would not do that as this could be error prone.

**Comment by ktf** [comment]
@davidrohr @shahor02 I think this could be merged as is. It is a step forward the current implementation and if the Lifetime::Condition is not used, there should not be any side effects at all. What do you think? The FLP suite will use in any case the tag of yesterday, no?

**Comment by shahor02** [comment]
Hi @ktf 
thanks, fine with me to merge this. I'll try to play with it after finishing another branch I am working with.

**Comment by shahor02** [comment]
@ktf were not you going to merge this PR? Is there any reason to not do this?

**Comment by davidrohr** [comment]
I discussed with @ktf to check this in FST with DD before merging, this test is currently running. Will comment here soon.

**Review by davidrohr — APPROVED** [review]
Don't see any problem in FST, fine for me to merge this.

**Comment by ktf** [comment]
Ok, thanks. As discussed, let's take it from here for the remaining bugs / features.

**Pull request merged**
